Startup Script reports `ok` too fast, and doesn't wait for actual start of Jetty

**Jetty version(s)**
10.0.16

**Jetty Environment**
N/A

**Java version/vendor** `(use: java -version)`
openjdk version "17.0.8" 2023-07-18 LTS

**OS type/version**
Amazon Linux 2

**Description**
Jetty starts fine, but startup script throws error. PID file is empty, so stopping the service is not working.


Startup:
```
# service jetty start
Starting Jetty: pgrep: option requires an argument -- 'P'

Usage:
 pgrep [options] <pattern>

Options:
 -d, --delimiter <string>  specify output delimiter
 -l, --list-name           list PID and process name
 -a, --list-full           list PID and full command line
 -v, --inverse             negates the matching
 -w, --lightweight         list all TID
 -c, --count               count of matching processes
 -f, --full                use full process name to match
 -g, --pgroup <PGID,...>   match listed process group IDs
 -G, --group <GID,...>     match real group IDs
 -n, --newest              select most recently started
 -o, --oldest              select least recently started
 -P, --parent <PPID,...>   match only child processes of the given parent
 -s, --session <SID,...>   match session IDs
 -t, --terminal <tty,...>  match by controlling terminal
 -u, --euid <ID,...>       match by effective IDs
 -U, --uid <ID,...>        match by real IDs
 -x, --exact               match exactly with the command name
 -F, --pidfile <file>      read PIDs from file
 -L, --logpidfile          fail if PID file is not locked
 --ns <PID>                match the processes that belong to the same
                           namespace as <pid>
 --nslist <ns,...>         list which namespaces will be considered for
                           the --ns option.
                           Available namespaces: ipc, mnt, net, pid, user, uts

 -h, --help     display this help and exit
 -V, --version  output version information and exit

For more details see pgrep(1).
ok Mon Sep  4 10:25:13 CEST 2023
```

Stop:
```
# service jetty stop
Stopping Jetty: ERROR: no pid id found in /var/jetty/run/jetty-helse.pid
```


**How to reproduce?**
Start and stop jetty

@l-cdc I intentionally setup an environment which will trigger permissions issues with `PidFile`.

It will result in the following output, and will fail to start Jetty.
So I feel we have this case covered, as long as the `PidFile` option actually attempts to run.

```
Starting Jetty: ok Wed Sep 13 15:01:32 UTC 2023
2023-09-13 15:01:32.399:WARN :oejx.XmlConfiguration:main: Config error java.lang.reflect.InvocationTargetException at <Call class="org.eclipse.jetty.util.PidFile" name="create"><Arg name="file"><Property name="jetty.pid" default="jetty.pid"/></Arg></Call> in file:///var/www/test/jetty-home-10.0.16/etc/jetty-pid.xml
2023-09-13 15:01:32.403:WARN :oejx.XmlConfiguration:main: Unable to execute XmlConfiguration
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeMethod(XmlConfiguration.java:777)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:985)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:950)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:504)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:457)
	at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:359)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1900)
Caused by: 
java.nio.file.AccessDeniedException: /var/www/test/jetty.pid
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
	at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
	at java.base/java.nio.file.Files.newOutputStream(Files.java:220)
	at java.base/java.nio.file.Files.write(Files.java:3425)
	at java.base/java.nio.file.Files.writeString(Files.java:3641)
	at org.eclipse.jetty.util.PidFile.create(PidFile.java:58)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeMethod(XmlConfiguration.java:777)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:985)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:950)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:504)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:457)
	at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:359)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1900)
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeMethod(XmlConfiguration.java:777)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:985)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:950)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:504)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:457)
	at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:359)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1900)
Caused by: java.nio.file.AccessDeniedException: /var/www/test/jetty.pid
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
	at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
	at java.base/java.nio.file.Files.newOutputStream(Files.java:220)
	at java.base/java.nio.file.Files.write(Files.java:3425)
	at java.base/java.nio.file.Files.writeString(Files.java:3641)
	at org.eclipse.jetty.util.PidFile.create(PidFile.java:58)
	... 11 more
```

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -140,25 +140,24 @@ started()
   # wait till timeout to see "STARTED" in state file, needs --module=state as argument
   for ((T = 0; T < $STARTTIMEOUT; T++))
   do
+    echo -n "."
     sleep 1
     if [ -r $STATEFILE ] ; then
       STATENOW=$(tail -1 $STATEFILE)
       (( DEBUG )) && echo "State (now): $STATENOW"
       case "$STATENOW" in
         STARTED*)
-          echo "."
+          echo " started"
           return 0;;
         STOPPED*)
-          echo "!"
+          echo " stopped"
           return 1;;
         FAILED*)
-          echo "!!"
+          echo " failed"
           return 1;;
       esac
-      echo -n "-"
     else
       (( DEBUG )) && echo "Unable to read State File: $STATEFILE"
-      echo -n ":"
     fi
   done
   (( DEBUG )) && echo "Timeout $STARTTIMEOUT expired waiting for start state from $STATEFILE"

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -136,16 +136,36 @@ started()
   STATEFILE=$1
   PIDFILE=$2
   STARTTIMEOUT=$3
-  # wait for 60s to see "STARTED" in state file, needs --module=state as argument
+  # wait till timeout to see "STARTED" in state file, needs --module=state as argument
   for ((T = 0; T < $(($STARTTIMEOUT / 4)); T++))
   do
     sleep 4
-    [ -z "$(tail -1 $STATEFILE | grep STARTED 2>/dev/null)" ] || return 0
-    [ -z "$(tail -1 $STATEFILE | grep STOPPED 2>/dev/null)" ] || return 1
-    [ -z "$(tail -1 $STATEFILE | grep FAILED 2>/dev/null)" ] || return 1
-    local PID=$(cat "$PIDFILE" 2>/dev/null) || return 1
-    kill -0 "$PID" 2>/dev/null || return 1
-    echo -n ". "
+    if [ -r $STATEFILE ]
+    then
+      STATENOW=$(tail -1 $STATEFILE)
+      (( DEBUG )) && echo "State (now): $STATENOW"
+      case "$STATENOW" in
+        STARTED*)
+          return 0;;
+        STOPPED*)
+          return 1;;
+        FAILED*)
+          return 1;;
+      esac
+      echo -n "-"
+    else
+      (( DEBUG )) && echo "Unable to read State File: $STATEFILE"
+      echo -n ":"
+    fi
+    if [ -r $PIDFILE ]
+    then
+      local PID=$(cat "$PIDFILE" 2> /dev/null) || return 1
+      kill -0 "$PID" 2> /dev/null || return 1
+      echo -n ". "
+    else
+      (( DEBUG )) && echo "Unable to read PID File: $PIDFILE"
+      echo -n "!"
+    fi
   done
 
   return 1;
@@ -440,10 +460,6 @@ esac
 JETTY_DRY_RUN=$("$JAVA" -jar "$JETTY_START" --dry-run=opts,path,main,args ${JETTY_ARGS[*]} ${JAVA_OPTIONS[*]})
 RUN_ARGS=($JETTY_SYS_PROPS ${JETTY_DRY_RUN[@]})
 
-#####################################################
-# Comment these out after you're happy with what
-# the script is doing.
-#####################################################
 if (( DEBUG ))
 then
   dumpEnv

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -132,7 +132,7 @@ running()
 
 started()
 {
-  # wait for 60s to see "STARTED" in PID file, needs jetty-started.xml as argument
+  # wait for 60s to see "STARTED" in PID file, needs jetty-state.xml as argument
   for ((T = 0; T < $(($3 / 4)); T++))
   do
     sleep 4
@@ -506,7 +506,7 @@ case "$ACTION" in
 
     fi
 
-    if expr "${JETTY_ARGS[*]}" : '.*jetty-started.xml.*' >/dev/null
+    if expr "${JETTY_ARGS[*]}" : '.*jetty-state.xml.*' >/dev/null
     then
       if started "$JETTY_STATE" "$JETTY_PID" "$JETTY_START_TIMEOUT"
       then

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -138,9 +138,9 @@ started()
   local PIDFILE=$2
   local STARTTIMEOUT=$3
   # wait till timeout to see "STARTED" in state file, needs --module=state as argument
-  for ((T = 0; T < $(($STARTTIMEOUT / 4)); T++))
+  for ((T = 0; T < $STARTTIMEOUT; T++))
   do
-    sleep 4
+    sleep 1
     if [ -r $STATEFILE ] ; then
       STATENOW=$(tail -1 $STATEFILE)
       (( DEBUG )) && echo "State (now): $STATENOW"
@@ -189,8 +189,10 @@ pidKill()
         (( DEBUG )) && echo "PID=$PID is running, sending kill signal=KILL (TIMEOUT=$TIMEOUT)"
         kill -KILL "$PID" 2>/dev/null
       fi
+      echo -n "."
       sleep 1
     done
+    echo "Killed $PID"
     return 0
   else
     (( DEBUG )) && echo "Unable to read PID File: $PIDFILE"

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -497,11 +497,11 @@ case "$ACTION" in
         su - "$JETTY_USER" $SU_SHELL -c "
           cd \"$JETTY_BASE\"
           echo ${RUN_ARGS[*]} start-log-file=\"$JETTY_START_LOG\" | xargs ${JAVA} > /dev/null &
-          disown $(pgrep -P $!)"
+          disown \$!"
       else
         # Startup if not switching users
         echo ${RUN_ARGS[*]} | xargs ${JAVA} > /dev/null &
-        disown $(pgrep -P $!)
+        disown $!
       fi
 
     fi

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -130,16 +130,20 @@ running()
   return 1
 }
 
+# Test state file (after timeout) for started state
 started()
 {
-  # wait for 60s to see "STARTED" in PID file, needs jetty-state.xml as argument
-  for ((T = 0; T < $(($3 / 4)); T++))
+  STATEFILE=$1
+  PIDFILE=$2
+  STARTTIMEOUT=$3
+  # wait for 60s to see "STARTED" in state file, needs --module=state as argument
+  for ((T = 0; T < $(($STARTTIMEOUT / 4)); T++))
   do
     sleep 4
-    [ -z "$(tail -1 $1 | grep STARTED 2>/dev/null)" ] || return 0
-    [ -z "$(tail -1 $1 | grep STOPPED 2>/dev/null)" ] || return 1
-    [ -z "$(tail -1 $1 | grep FAILED 2>/dev/null)" ] || return 1
-    local PID=$(cat "$2" 2>/dev/null) || return 1
+    [ -z "$(tail -1 $STATEFILE | grep STARTED 2>/dev/null)" ] || return 0
+    [ -z "$(tail -1 $STATEFILE | grep STOPPED 2>/dev/null)" ] || return 1
+    [ -z "$(tail -1 $STATEFILE | grep FAILED 2>/dev/null)" ] || return 1
+    local PID=$(cat "$PIDFILE" 2>/dev/null) || return 1
     kill -0 "$PID" 2>/dev/null || return 1
     echo -n ". "
   done
@@ -506,7 +510,7 @@ case "$ACTION" in
 
     fi
 
-    if expr "${JETTY_ARGS[*]}" : '.*jetty-state.xml.*' >/dev/null
+    if expr "${JETTY_ARGS[*]}" : '.*jetty\.state=.*' >/dev/null
     then
       if started "$JETTY_STATE" "$JETTY_PID" "$JETTY_START_TIMEOUT"
       then

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -118,30 +118,30 @@ findDirectory()
   done
 }
 
+# test if process specified in PID file is still running
 running()
 {
-  if [ -f "$1" ]
-  then
-    local PID=$(cat "$1" 2>/dev/null) || return 1
-    kill -0 "$PID" 2>/dev/null
-    return
+  local PIDFILE=$1
+  if [ -r "$PIDFILE" ] ; then
+    local PID=$(tail -1 "$PIDFILE")
+    if kill -0 "$PID" 2>/dev/null ; then
+      return 0
+    fi
   fi
-  rm -f "$1"
   return 1
 }
 
 # Test state file (after timeout) for started state
 started()
 {
-  STATEFILE=$1
-  PIDFILE=$2
-  STARTTIMEOUT=$3
+  local STATEFILE=$1
+  local PIDFILE=$2
+  local STARTTIMEOUT=$3
   # wait till timeout to see "STARTED" in state file, needs --module=state as argument
   for ((T = 0; T < $(($STARTTIMEOUT / 4)); T++))
   do
     sleep 4
-    if [ -r $STATEFILE ]
-    then
+    if [ -r $STATEFILE ] ; then
       STATENOW=$(tail -1 $STATEFILE)
       (( DEBUG )) && echo "State (now): $STATENOW"
       case "$STATENOW" in
@@ -157,18 +157,45 @@ started()
       (( DEBUG )) && echo "Unable to read State File: $STATEFILE"
       echo -n ":"
     fi
-    if [ -r $PIDFILE ]
-    then
-      local PID=$(cat "$PIDFILE" 2> /dev/null) || return 1
-      kill -0 "$PID" 2> /dev/null || return 1
-      echo -n ". "
-    else
-      (( DEBUG )) && echo "Unable to read PID File: $PIDFILE"
-      echo -n "!"
-    fi
   done
-
   return 1;
+}
+
+pidKill()
+{
+  local PIDFILE=$1
+  local TIMEOUT=$2
+
+  if [ -r $PIDFILE ] ; then
+    local PID=$(tail -1 "$PIDFILE")
+    if [ -z "$PID" ] ; then
+      echo "ERROR: no pid found in $PIDFILE"
+      return 1
+    fi
+
+    # Try default kill first
+    if kill -0 "$PID" 2>/dev/null ; then
+      (( DEBUG )) && echo "PID=$PID is running, sending kill"
+      kill "$PID" 2>/dev/null
+    else
+      rm -f $PIDFILE 2> /dev/null
+      return 0
+    fi
+
+    # Perform harsh kill next
+    while kill -0 "$PID" 2>/dev/null
+    do
+      if (( TIMEOUT-- == 0 )) ; then
+        (( DEBUG )) && echo "PID=$PID is running, sending kill signal=KILL (TIMEOUT=$TIMEOUT)"
+        kill -KILL "$PID" 2>/dev/null
+      fi
+      sleep 1
+    done
+    return 0
+  else
+    (( DEBUG )) && echo "Unable to read PID File: $PIDFILE"
+    return 1
+  fi
 }
 
 
@@ -533,6 +560,7 @@ case "$ACTION" in
         echo "OK `date`"
       else
         echo "FAILED `date`"
+        pidKill $JETTY_PID 30
         exit 1
       fi
     else
@@ -557,26 +585,12 @@ case "$ACTION" in
       done
     else
       # Stop from a non-service path
-      if [ ! -f "$JETTY_PID" ] ; then
+      if [ ! -r "$JETTY_PID" ] ; then
         echo "ERROR: no pid found at $JETTY_PID"
         exit 1
       fi
 
-      PID=$(cat "$JETTY_PID" 2>/dev/null)
-      if [ -z "$PID" ] ; then
-        echo "ERROR: no pid id found in $JETTY_PID"
-        exit 1
-      fi
-      kill "$PID" 2>/dev/null
-
-      TIMEOUT=30
-      while running $JETTY_PID; do
-        if (( TIMEOUT-- == 0 )); then
-          kill -KILL "$PID" 2>/dev/null
-        fi
-
-        sleep 1
-      done
+      pidKill "$JETTY_PID" 30
     fi
 
     rm -f "$JETTY_PID"

--- a/jetty-home/src/main/resources/bin/jetty.sh
+++ b/jetty-home/src/main/resources/bin/jetty.sh
@@ -146,10 +146,13 @@ started()
       (( DEBUG )) && echo "State (now): $STATENOW"
       case "$STATENOW" in
         STARTED*)
+          echo "."
           return 0;;
         STOPPED*)
+          echo "!"
           return 1;;
         FAILED*)
+          echo "!!"
           return 1;;
       esac
       echo -n "-"
@@ -158,6 +161,8 @@ started()
       echo -n ":"
     fi
   done
+  (( DEBUG )) && echo "Timeout $STARTTIMEOUT expired waiting for start state from $STATEFILE"
+  echo " timeout"
   return 1;
 }
 

--- a/jetty-server/src/main/config/modules/state.mod
+++ b/jetty-server/src/main/config/modules/state.mod
@@ -9,6 +9,9 @@ start
 [depends]
 server
 
+[before]
+deploy
+
 [xml]
 etc/jetty-state.xml
 

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
@@ -32,7 +32,7 @@ import static java.nio.file.StandardOpenOption.WRITE;
  * A LifeCycle Listener that writes state changes to a file.
  * <p>This can be used with the jetty.sh script to wait for successful startup.
  */
-public class StateLifeCycleListener extends Thread implements LifeCycle.Listener
+public class StateLifeCycleListener implements LifeCycle.Listener
 {
     private static final Logger LOG = LoggerFactory.getLogger(StateLifeCycleListener.class);
 
@@ -55,22 +55,6 @@ public class StateLifeCycleListener extends Thread implements LifeCycle.Listener
 
         // Create file
         Files.writeString(stateFile, "INIT " + this + "\n", UTF_8, WRITE, CREATE_NEW);
-
-        // JVM shutdown should clean up state file
-        Runtime.getRuntime().addShutdownHook(this);
-    }
-
-    @Override
-    public void run()
-    {
-        try
-        {
-            Files.deleteIfExists(stateFile);
-        }
-        catch (Throwable t)
-        {
-            LOG.info("Unable to remove State file: {}", stateFile, t);
-        }
     }
 
     private void appendStateChange(String action, Object obj)

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
@@ -13,6 +13,7 @@
 
 package org.eclipse.jetty.server;
 
+import java.io.IOException;
 import java.io.Writer;
 import java.nio.file.Files;
 import java.nio.file.Path;
@@ -24,7 +25,7 @@ import org.slf4j.LoggerFactory;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.nio.file.StandardOpenOption.APPEND;
-import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
 import static java.nio.file.StandardOpenOption.WRITE;
 
 /**
@@ -37,53 +38,75 @@ public class StateLifeCycleListener implements LifeCycle.Listener
 
     private final Path _filename;
 
-    public StateLifeCycleListener(String filename)
+    public StateLifeCycleListener(String filename) throws IOException
     {
         _filename = Paths.get(filename).toAbsolutePath();
 
         if (LOG.isDebugEnabled())
             LOG.debug("State File: {}", _filename);
+
+        // We use raw Files APIs here to allow important IOExceptions
+        // to fail the startup of Jetty, as these kinds of errors
+        // point to filesystem permission issues that must be resolved
+        // by the user before the state file can be used.
+
+        // Start with fresh file (for permission reasons)
+        Files.deleteIfExists(_filename);
+
+        // Create file
+        Files.writeString(_filename, "INIT " + this + "\n", UTF_8, WRITE, CREATE_NEW);
     }
 
-    private void writeState(String action, LifeCycle lifecycle)
+    @Override
+    public String toString()
     {
-        try (Writer out = Files.newBufferedWriter(_filename, UTF_8, WRITE, CREATE, APPEND))
+        return String.format("%s@%h", this.getClass().getSimpleName(), this);
+    }
+
+    private void appendStateChange(String action, Object obj)
+    {
+        try (Writer out = Files.newBufferedWriter(_filename, UTF_8, WRITE, APPEND))
         {
-            out.append(action).append(" ").append(lifecycle.toString()).append("\n");
+            String entry = String.format("%s %s\n", action, obj);
+            if (LOG.isDebugEnabled())
+                LOG.debug("appendEntry to {}: {}", _filename, entry);
+            out.append(entry);
         }
-        catch (Exception e)
+        catch (IOException e)
         {
-            LOG.warn("Unable to write state", e);
+            // this can happen if the uid of the Jetty process changes after it has been started
+            // such as can happen with some setuid configurations
+            LOG.warn("Unable to append to state file: " + _filename, e);
         }
     }
 
     @Override
     public void lifeCycleStarting(LifeCycle event)
     {
-        writeState("STARTING", event);
+        appendStateChange("STARTING", event);
     }
 
     @Override
     public void lifeCycleStarted(LifeCycle event)
     {
-        writeState("STARTED", event);
+        appendStateChange("STARTED", event);
     }
 
     @Override
     public void lifeCycleFailure(LifeCycle event, Throwable cause)
     {
-        writeState("FAILED", event);
+        appendStateChange("FAILED", event);
     }
 
     @Override
     public void lifeCycleStopping(LifeCycle event)
     {
-        writeState("STOPPING", event);
+        appendStateChange("STOPPING", event);
     }
 
     @Override
     public void lifeCycleStopped(LifeCycle event)
     {
-        writeState("STOPPED", event);
+        appendStateChange("STOPPED", event);
     }
 }

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
@@ -57,12 +57,6 @@ public class StateLifeCycleListener implements LifeCycle.Listener
         Files.writeString(_filename, "INIT " + this + "\n", UTF_8, WRITE, CREATE_NEW);
     }
 
-    @Override
-    public String toString()
-    {
-        return String.format("%s@%h", this.getClass().getSimpleName(), this);
-    }
-
     private void appendStateChange(String action, Object obj)
     {
         try (Writer out = Files.newBufferedWriter(_filename, UTF_8, WRITE, APPEND))
@@ -108,5 +102,11 @@ public class StateLifeCycleListener implements LifeCycle.Listener
     public void lifeCycleStopped(LifeCycle event)
     {
         appendStateChange("STOPPED", event);
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("%s@%h", this.getClass().getSimpleName(), this);
     }
 }

--- a/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
+++ b/jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java
@@ -32,18 +32,18 @@ import static java.nio.file.StandardOpenOption.WRITE;
  * A LifeCycle Listener that writes state changes to a file.
  * <p>This can be used with the jetty.sh script to wait for successful startup.
  */
-public class StateLifeCycleListener implements LifeCycle.Listener
+public class StateLifeCycleListener extends Thread implements LifeCycle.Listener
 {
     private static final Logger LOG = LoggerFactory.getLogger(StateLifeCycleListener.class);
 
-    private final Path _filename;
+    private final Path stateFile;
 
     public StateLifeCycleListener(String filename) throws IOException
     {
-        _filename = Paths.get(filename).toAbsolutePath();
+        stateFile = Paths.get(filename).toAbsolutePath();
 
         if (LOG.isDebugEnabled())
-            LOG.debug("State File: {}", _filename);
+            LOG.debug("State File: {}", stateFile);
 
         // We use raw Files APIs here to allow important IOExceptions
         // to fail the startup of Jetty, as these kinds of errors
@@ -51,26 +51,42 @@ public class StateLifeCycleListener implements LifeCycle.Listener
         // by the user before the state file can be used.
 
         // Start with fresh file (for permission reasons)
-        Files.deleteIfExists(_filename);
+        Files.deleteIfExists(stateFile);
 
         // Create file
-        Files.writeString(_filename, "INIT " + this + "\n", UTF_8, WRITE, CREATE_NEW);
+        Files.writeString(stateFile, "INIT " + this + "\n", UTF_8, WRITE, CREATE_NEW);
+
+        // JVM shutdown should clean up state file
+        Runtime.getRuntime().addShutdownHook(this);
+    }
+
+    @Override
+    public void run()
+    {
+        try
+        {
+            Files.deleteIfExists(stateFile);
+        }
+        catch (Throwable t)
+        {
+            LOG.info("Unable to remove State file: {}", stateFile, t);
+        }
     }
 
     private void appendStateChange(String action, Object obj)
     {
-        try (Writer out = Files.newBufferedWriter(_filename, UTF_8, WRITE, APPEND))
+        try (Writer out = Files.newBufferedWriter(stateFile, UTF_8, WRITE, APPEND))
         {
             String entry = String.format("%s %s\n", action, obj);
             if (LOG.isDebugEnabled())
-                LOG.debug("appendEntry to {}: {}", _filename, entry);
+                LOG.debug("appendEntry to {}: {}", stateFile, entry);
             out.append(entry);
         }
         catch (IOException e)
         {
             // this can happen if the uid of the Jetty process changes after it has been started
             // such as can happen with some setuid configurations
-            LOG.warn("Unable to append to state file: " + _filename, e);
+            LOG.warn("Unable to append to state file: " + stateFile, e);
         }
     }
 

--- a/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java
+++ b/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java
@@ -156,6 +156,7 @@ public class DistributionTests extends AbstractJettyHomeTest
             }
 
             await().atMost(Duration.ofSeconds(10)).until(() -> !Files.exists(pidfile));
+            await().atMost(Duration.ofSeconds(10)).until(() -> !Files.exists(statefile));
         }
     }
 

--- a/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java
+++ b/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java
@@ -16,6 +16,7 @@ package org.eclipse.jetty.tests.distribution;
 import java.io.BufferedWriter;
 import java.io.File;
 import java.io.FileWriter;
+import java.io.IOException;
 import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -74,6 +75,7 @@ import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.startsWith;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNotEquals;
@@ -136,7 +138,7 @@ public class DistributionTests extends AbstractJettyHomeTest
             args.add("jetty.pid=" + pidfile);
 
             Path confFile = run1.getConfig().getJettyHome().resolve("etc/jetty.conf");
-            for (String line: Files.readAllLines(confFile, StandardCharsets.UTF_8))
+            for (String line : Files.readAllLines(confFile, StandardCharsets.UTF_8))
             {
                 if (line.startsWith("#") || StringUtil.isBlank(line))
                     continue; // skip
@@ -149,6 +151,8 @@ public class DistributionTests extends AbstractJettyHomeTest
 
                 assertTrue(Files.isRegularFile(pidfile), "PID file should exist");
                 assertTrue(Files.isRegularFile(statefile), "State file should exist");
+                String state = tail(statefile);
+                assertThat("State file", state, startsWith("STARTED "));
 
                 startHttpClient();
                 ContentResponse response = client.GET("http://localhost:" + port);
@@ -156,7 +160,28 @@ public class DistributionTests extends AbstractJettyHomeTest
             }
 
             await().atMost(Duration.ofSeconds(10)).until(() -> !Files.exists(pidfile));
-            await().atMost(Duration.ofSeconds(10)).until(() -> !Files.exists(statefile));
+            await().atMost(Duration.ofSeconds(10)).until(() -> tail(statefile).startsWith("STOPPED "));
+        }
+    }
+
+    /**
+     * Get the last line of the file.
+     *
+     * @param file the file to read from
+     * @return the string representing the last line of the file, or null if not found
+     */
+    private static String tail(Path file)
+    {
+        try
+        {
+            List<String> lines = Files.readAllLines(file, StandardCharsets.UTF_8);
+            if (lines.isEmpty())
+                return "";
+            return lines.get(lines.size() - 1);
+        }
+        catch (IOException e)
+        {
+            return "";
         }
     }
 
@@ -1321,7 +1346,6 @@ public class DistributionTests extends AbstractJettyHomeTest
                 assertThat(response.getStatus(), is(HttpStatus.OK_200));
                 content = response.getContentAsString();
                 assertThat(content, containsString("not authenticated"));
-
             }
         }
         finally
@@ -1359,7 +1383,7 @@ public class DistributionTests extends AbstractJettyHomeTest
             int port = distribution.freePort();
             String[] args2 = {
                 "jetty.http.port=" + port,
-            };
+                };
             try (JettyHomeTester.Run run2 = distribution.start(args2))
             {
                 assertTrue(run2.awaitConsoleLogsFor("Started Server@", 10, TimeUnit.SECONDS));
@@ -1375,7 +1399,7 @@ public class DistributionTests extends AbstractJettyHomeTest
 
                 Path requestLog = distribution.getJettyBase().resolve("logs/test.request.log");
                 List<String> loggedLines = Files.readAllLines(requestLog, StandardCharsets.UTF_8);
-                for (String loggedLine: loggedLines)
+                for (String loggedLine : loggedLines)
                 {
                     assertThat(loggedLine, containsString(" [foo space here] "));
                 }

--- a/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java
+++ b/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java
@@ -24,6 +24,7 @@ import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
 import java.nio.file.StandardOpenOption;
 import java.time.Duration;
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Queue;
@@ -49,6 +50,7 @@ import org.eclipse.jetty.unixsocket.client.HttpClientTransportOverUnixSockets;
 import org.eclipse.jetty.unixsocket.server.UnixSocketConnector;
 import org.eclipse.jetty.util.BlockingArrayQueue;
 import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.StringUtil;
 import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.websocket.api.Session;
 import org.eclipse.jetty.websocket.api.StatusCode;
@@ -106,6 +108,54 @@ public class DistributionTests extends AbstractJettyHomeTest
                 run2.stop();
                 assertTrue(run2.awaitFor(10, TimeUnit.SECONDS));
             }
+        }
+    }
+
+    @Test
+    public void testJettyConf() throws Exception
+    {
+        String jettyVersion = System.getProperty("jettyVersion");
+        JettyHomeTester distribution = JettyHomeTester.Builder.newInstance()
+            .jettyVersion(jettyVersion)
+            .mavenLocalRepository(System.getProperty("mavenRepoPath"))
+            .build();
+
+        try (JettyHomeTester.Run run1 = distribution.start("--add-modules=http"))
+        {
+            assertTrue(run1.awaitFor(10, TimeUnit.SECONDS));
+            assertEquals(0, run1.getExitValue());
+
+            Path pidfile = run1.getConfig().getJettyBase().resolve("jetty.pid");
+            Path statefile = run1.getConfig().getJettyBase().resolve("jetty.state");
+
+            int port = distribution.freePort();
+
+            List<String> args = new ArrayList<>();
+            args.add("jetty.http.port=" + port);
+            args.add("jetty.state=" + statefile);
+            args.add("jetty.pid=" + pidfile);
+
+            Path confFile = run1.getConfig().getJettyHome().resolve("etc/jetty.conf");
+            for (String line: Files.readAllLines(confFile, StandardCharsets.UTF_8))
+            {
+                if (line.startsWith("#") || StringUtil.isBlank(line))
+                    continue; // skip
+                args.add(line);
+            }
+
+            try (JettyHomeTester.Run run2 = distribution.start(args))
+            {
+                assertTrue(run2.awaitConsoleLogsFor("Started Server@", 10, TimeUnit.SECONDS));
+
+                assertTrue(Files.isRegularFile(pidfile), "PID file should exist");
+                assertTrue(Files.isRegularFile(statefile), "State file should exist");
+
+                startHttpClient();
+                ContentResponse response = client.GET("http://localhost:" + port);
+                assertEquals(HttpStatus.NOT_FOUND_404, response.getStatus());
+            }
+
+            await().atMost(Duration.ofSeconds(10)).until(() -> !Files.exists(pidfile));
         }
     }
 

